server/src/main/java/org/elasticsearch/search/crossproject/CrossProjectIndexResolutionValidator.java

```diff
@@ -100,85 +112,167 @@ public static ElasticsearchException validate(
+                        remoteUnauthorizedIndices = new HashMap<>();
+                    }
+                    populateRemoteSecurityExceptionAndIndices.accept(remoteAuthorizationExceptions, remoteUnauthorizedIndices);
                 }
-                // Prefer reporting 404 on origin over linked projects
-                if (localException != null) {
+                if (currentExpressionSecurityException != null && currentExpressionSecurityException == localException) {
+                    // We have a local security exception for this unqualified expression. That's what we want to report, i.e.
+                    // we are no longer interested in whether any linked project also returns a security exception.
+                    assert populateRemoteSecurityExceptionAndIndices == null;
+                    if (localAuthorizationException == null) {
+                        localAuthorizationException = currentExpressionSecurityException;
+                        localUnauthorizedIndices = new ArrayList<>();
+                    }
+                    localUnauthorizedIndices.add(originalExpression);
+                } else if (localException != null) {
+                    // We have a local 404 for this unqualified expression which takes priority over any remote 404
                     assert localException instanceof IndexNotFoundException
                         : "Expected local exception to be IndexNotFoundException, but found: " + localException;
-                    return localException;
+                    if (notFoundException == null) {
+                        notFoundException = (IndexNotFoundException) localException;
+                    }
                 } else {
+                    // Local project is excluded and 404 from all remotes, we report 404 for the first remote project
+                    assert localExpressions == ResolvedIndexExpression.LocalExpressions.NONE : localExpressions;
                     assert false == remoteExpressions.isEmpty() : "expected remote expressions to be non-empty";
-                    return new IndexNotFoundException(remoteExpressions.iterator().next());
+                    if (notFoundException == null) {
+                        notFoundException = new IndexNotFoundException(remoteExpressions.getFirst());
+                    }
                 }
             }
         }
-        // if we didn't throw before it means that we can proceed with the request
-        return null;
+
+        if (localAuthorizationException == null && remoteAuthorizationExceptions == null) {
+            return notFoundException;
+        } else {
+            var firstException = localAuthorizationException != null
+                ? formatAuthorizationException(localAuthorizationException, localUnauthorizedIndices)
+                : null;
+
+            if (remoteAuthorizationExceptions != null) {
+                for (var e : remoteAuthorizationExceptions.entrySet()) {
```

Comment:
Iterating through this `entrySet()` can still be non-deterministic. I think we should either sort it here or use  `LinkedHashMap` for `remoteAuthorizationExceptions`. I prefer the later.